### ISSUE TITLE
Caching fallback breaks when external cache system has problems

### ISSUE THREAD

**Issue by yuliu** [mybb/mybb#3793]
Community [post](https://community.mybb.com/thread-224537.html). I've discussed much with @dvz on this issue, and I'll post it here. It's a little bit long, but I think it's worth reading, especially for those who are not familiar with how MyBB's caching works. Sorry, no TL;DR this time. 😆 

I think it's an issue that really breaks whole MyBB's running, eg. forums/forumpermissions/usergroups/etc.

Well, I did some research on MyBB's caching mechanism which supports using database and external cache system as cache handler, and will fallback to using database if the cache handler fails. Let me assume some failures that can really happen when using an external cache system, taking Memcache (I've tested with it for sure, probably Memcached too) for example:

1. If MyBB's cache setting is correct:
- The cache system, eg. Memcache, is gone/malfunctioning/etc. MyBB can't fetch cache from it.
- The cache system is restarted, but currently no cache exists. MyBB can fetch cache but will get none, then MyBB tries to create cache on the cache system.

2. If MyBB's cache setting is wrong with Memcache settings, eg `host` and `port`. MyBB can't fetch cache from it.

It's worth noting that when an external cache system is invalid or MyBB gets invalid cache from it (probably a `bool(false)`), MyBB tries to fetch the cache content stored in the database, which is a fallback mechanism, and then recreate the cache on external cache system (if the cache system's still working).
For the second situation, MyBB will not test the cache server's connectivity before trying to fetch cache from it, yielding MyBB gets invalid cache from it, [a `bool(false)` as documented by PHP](https://www.php.net/manual/en/memcache.get.php#refsect1-memcache.get-returnvalues).

The whole caching work should be working just fine if we have a valid data-cache configuration (in `./inc/config.php`). But for large forum, whatever with a large number of categories/subforums or a large number of user groups or something else, the cache fallback mechanism will be broken.

Internally (and externally in a very few modules), MyBB uses a custom (un)serialize function to store and recover some internal data, for this issue, it's cache content. The corresponding functions of PHP's built-in ones are [`serialize()`](https://www.php.net/manual/en/function.serialize.php) and [`unserialize()`](https://www.php.net/manual/en/function.unserialize.php). For security concerns, MyBB changes to use custom (un)serialize functions instead, i.e., `my_(un)serialize()`, see #1733.

However, the `_safe_unserialize()` function used by `my_unserialize()` has a hard limit for the length of input string to be unserialized. When the length exceeds this limit, the function just refuses to work and returns a `bool(false)`. This is indeed a problem/limitation, see also #3775, for related issues.

Let me get back to the cache problem. When MyBB gets a `bool(false)` from the external cache system, it'll fall back to query on database (table `datacache`) for that value. When MyBB encounters a long input string, which is read from a record in table `datacache` to be unserialized, the custom `my_unserialize()` will return a `bool(false)` to the fallback query. Yes, it's the final answer, any failed unserialization will lead to a cache to be invalid with a `bool(false)` value that is used everywhere where the cache is read.

For the community post, OP observed this problem when his Memcached server is restart, meaning it's working, but MyBB only gets `false` upon cache reading. Actually this `false` means the key doesn't exist. Then MyBB falls back on database query to fetch the original cache data, and fails, probably a very long field.

Well, the story thing can be stopped now. Here's some thoughts and resolutions to the problem. Not really well-designed for all but only a mitigation and something to be expected in future version:

- A quick fix for OP's problem is replace `my_unserialize` with `unserialize` in the [datacache class](https://github.com/mybb/mybb/blob/fda816dfab9d93c24f95ba76a925ac6bde72b3c4/inc/class_datacache.php#L166). See also what the cache reading is done using database as the cache store in [this line](https://github.com/mybb/mybb/blob/fda816dfab9d93c24f95ba76a925ac6bde72b3c4/inc/class_datacache.php#L195). This should solve most of the problems, including following concerns.

- Using an external cache system (sorry, I just realized really should use `cache store` instead `cache system` 🐞), to specific, Memcache/Memcached, MyBB doesn't check the connectivity to the cache server. This could be because the need of using clustering. However, the [Memcache::get](https://www.php.net/manual/en/memcache.get.php) during [`$cache->fetch()`](https://github.com/mybb/mybb/blob/fda816dfab9d93c24f95ba76a925ac6bde72b3c4/inc/cachehandlers/interface.php#L29) for Memcache/Memcached does throw a **notice** (not sure if it's E_NOTICE or E_USER_NOTICE ,it's even not a warning) if the connection from MyBB to the cache server is timeout (probably by a misconfiguration or a real network issue). MyBB suppresses notice by default, so the forum users (really should be owners/admins) are not aware of what really happens. Don't know if this kind of notice can be caught and alert forum admins.

- This one is a little off current topic. @dvz mentioned this when we were discussing about this issue. The `datacache` database table is volatile, meaning as its table name indicates, which is full of caches. As the circumstance for MyBB for now, almost all scripts of it rely on a global `$cache` variable which is built by reading from an external cache store or the `datacache` table. And also taking the cache fallback mechanism into account, this table should be really really really reliable. However currently, there lacks of a checking policy and rebuilding mechanism on existence/correctness of records in this table. There's still more, missing records if any will not be recreate by "Rebuild & Reload All" in AdminCP's Cache Manager.

- This one is a bit more off current topic. As I've mentioned, seems MyBB needs a **self** health checking mechanism, which is not just considered as a tool that requires user to click to run, on database and files integrity. The [DVZ Integrity Tools](https://community.mybb.com/mods.php?action=view&pid=980) is a great reference, but we might need some of its feature to run automatically like a task. Current design permits admins to delete/disable a task, so we need to add some fields to prevent MyBB itself to delete/disable health check tasks.

Oh.. seems pretty long is this post, but I think I've got the problem revealed clearly.

**Comment by euantorano** [comment]
There are a couple of other improvements I would like to see for caching:

- Add support for Redis as a cache driver
- Remove support for eaccelerator as a cache driver since it hasn't been updated since 2012: https://github.com/eaccelerator/eaccelerator

I'd like to replace the cache system completely in 1.9. Doing so, we could solve all of these issues in one easy swoop. I'd like to replace it with Laravel's cache library. The Laravel cache library has some neat functionality for retrieving values and falling back to a function to build the required data (and then cache it) if the cache doesn't exist:

```php
$value = Cache::get('key', function () {
    return DB::table(...)->get();
});
```

Full documentation for the library is here: https://laravel.com/docs/6.0/cache

**Comment by yuliu** [comment]
Yep. I was about to reference #3685 😆 , but haven't done that for regarding it a little bit off the problem this issue tries to address in current 1.8 version. Current cache system seems really have security issues, I'll try to post it in the private forum.

**Comment by yuliu** [comment]
Seems APC (2012, PHP 5.1+), eAccelerator (2012, PHP 5.1+) and XCache (2014, PHP 5.4+) are [all dead](https://en.wikipedia.org/wiki/List_of_PHP_accelerators#Compatibility_chart), and I even can't find detailed API documents for the latter two. The cache system in 1.9 should be re-organized/replaced.

@dvz I've implemented a cache check & recreated mechanism in #3801, we may have further discussion upon it.

**Comment by kawaii** [comment]
IIRC there is some discussion about adding a Redis cache driver to 1.9 or 1.10, so it would make sense for us to remove/deprecate older caches that are certainly no longer being used.

**Comment by euantorano** [comment]
APC is replaced by APCu: https://github.com/krakjoe/apcu

**Comment by yuliu** [comment]
A related but not exactly the same issue was reported in https://community.mybb.com/thread-229912.html :

OP described that their server is short on memory, MariaDB crashes often and memcached stayed up. During MariaDB down time, OP encountered the same thing as described in https://community.mybb.com/thread-224537.html , private forums lost their permission settings.

By the way, the [original issue](https://github.com/mybb/mybb/issues/3793#issue-490744872) appears to be hard to read and needs to be cleaned.

**Comment by lairdshaw** [comment]
The demand for a fix to the bug at the heart of this issue has been expressed at least twice in recent times in our Discord channel. The simple fix which @yuliu proposes in the first post in this issue - to "replace my_unserialize with unserialize in the [datacache class](https://github.com/mybb/mybb/blob/fda816dfab9d93c24f95ba76a925ac6bde72b3c4/inc/class_datacache.php#L166)" - seems to me, after a very recent Discord discussion starting [here](https://discordapp.com/channels/215876847634743296/316228730680770570/1029753842054746222), to be appropriate and without drawback. My proposal then is that the custom `my_(un)serialize()` functions, and those around which they wrap, `_safe_(un)serialize()`, be removed from the codebase, and calls to them replaced with calls to the standard PHP `(un)serialize()` functions. If anybody knows of a good reason not to do this, then please let us know. Otherwise, I think we should resolve this issue in this way ASAP, because it is a serious one, with an apparently very simple and effective solution.

**Comment by dvz** [comment]
The manual still advises against using PHP's functions for potentially untrusted input (even if trusted, we'd want to avoid the possibility of escalating cache/database access security problems into code execution).

It seems more acceptable to raise or remove the limits at
https://github.com/mybb/mybb/blob/c3c474a5b70f9fb4bb368b5b58c623b1c9e2fc16/inc/functions.php#L2395-L2400
and keep the scope of problems to performance only.

**Comment by lairdshaw** [comment]
I see that in the manual now. It's odd because they don't explain how, if `allowed_classes` is set false, there *is* potential for code to be loaded and run, but you're right that we need to take that warning seriously anyway.

The idea to change/remove the arbitrary limits had occurred to me too. My preference would be to remove them altogether. I haven't tested that but see no reason why it wouldn't work.

**Comment by dvz** [comment]
Tracking other contributing factors separately in #4615, #4616.
